### PR TITLE
feat(vscode): add support for "workbench.extensions.installExtension"

### PR DESCRIPTION
**Pull request by fgreinacher**
#### What it does

Adds support for the `workbench.extensions.installExtension` command (part of #8387).

We want this because we're considering to use https://github.com/joelspadin-garmin/vscode-private-extension-manager which depends on this API to install extensions.

#### How to test

1. Build the Example app
1. Put the [test extension](https://github.com/fgreinacher/Extension-Commands/releases/tag/3.0.0) in the `plugins` directory
1. Execute the command "Install extension via ID and VS Code API" with a valid an extension ID (e.g. `redhat.java`).
1. Execute the command "Install extension via VSIX and VS Code API" with the URI to a locally available VSIX file (something like `file:///home/gitpod/temp/some-other.vsix`)
1. Ensure that both extensions have been installed via the extension manager.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

### CODE REVIEW

File: packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts
Comment (fgreinacher) [inline packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:215]: VS Code accepts either an URI (for installing a plain vsix file) or a plain string (for installing by ID), see https://github.com/microsoft/vscode/blob/33a49a9c6328a4576f1d8692bee70c267cbfe16b/src/vs/workbench/contrib/extensions/browser/extensions.contribution.ts#L191

**Comment by fgreinacher** [comment]
Except some tests, am I missing anything important here?

**Review by vince-fugnitto — COMMENTED** [review]
@fgreinacher can you update the 'how to test' section of the pull-request so reviewers can easily review and test the functionality? Generally, in order to review changes to the plugin API, a corresponding plugin is provided or implemented to validate the feature.

**Comment by DukeNgn** [comment]
I created a quick test plugin to verify this PR. But the changes are not working on my side atm.

+ Download [test plugin](https://github.com/DukeNgn/Extension-Command/releases/download/v1.0/extcommand-0.0.1.vsix) ([source](https://github.com/DukeNgn/Extension-Commands)) and place in `plugins` folder in theia.
+ Restart the example application.
+ Call `Install Material Theme Icons (test installing extension by id)` from the command palette.


+ `Material Theme Icons` should be installed (it is not currently).

**Comment by fgreinacher** [comment]
Thanks a ton @DukeNgn!

I played around a bit. The problem is that the plugin server currently cannot deploy `.vsix` files. Deploying extensions by ID works when adding the `vscode:extension/` prefix, e.g. `vscode:extension/pkief.material-icon-theme".

File: packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts
Comment (fgreinacher) [inline packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:217]: Mirroring the logic from https://github.com/microsoft/vscode/blob/33a49a9c6328a4576f1d8692bee70c267cbfe16b/src/vs/workbench/contrib/extensions/browser/extensions.contribution.ts#L204.

File: packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts
Comment (fgreinacher) [inline packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts:27]: I'm not sure I like that, it's mostly a copy of `LocalDirectoryPluginDeployerResolver`. Maybe someone has an idea on how to improve that.

**Comment by fgreinacher** [comment]
> @fgreinacher can you update the 'how to test' section of the pull-request so reviewers can easily review and test the functionality? Generally, in order to review changes to the plugin API, a corresponding plugin is provided or implemented to validate the feature.

@vince-fugnitto I added some test steps, hope they are understandable! Looking forward to some feedback!

**Review by vince-fugnitto — COMMENTED** [review]
I believe that the following comment will need to be addressed: 
- https://github.com/eclipse-theia/theia/pull/8745/files#r523312500.

I'm however not sure about the proper approach, for example if we should merge `file` and `directory` resolvers.

File: packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts
Comment (benoitf) [inline packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts:32]: looks like a copy/paste ?

File: packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts
Comment (benoitf) [inline packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts:29]: should be only `file` no ?

file:///path-to-file

Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts:29]: That was also my first thought, but I decided to mirror `local-dir:`, which expects a filesystem path after the prefix. Using `file://` would for imply an URI for me, which this is currently not.

File: packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts
Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts:25]: I unified the common logic in ea032a21c83f7b062cf98b2b5f3f38811f5e0248.

**Comment by fgreinacher** [comment]
> I believe that the following comment will need to be addressed:
> 
> * https://github.com/eclipse-theia/theia/pull/8745/files#r523312500.
> 
> I'm however not sure about the proper approach, for example if we should merge `file` and `directory` resolvers.

I unified the common logic in ea032a21c83f7b062cf98b2b5f3f38811f5e0248.

File: packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts
Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts:27]: In addition, if we support the generic `file://` scheme I would expect that this works for files as well as directories

**Comment by azatsarynnyy** [comment]
@fgreinacher does this PR enables installing VS Code extensions by ID from Microsoft's Marketplace?

**Comment by fgreinacher** [comment]
> @fgreinacher does this PR enables installing VS Code extensions by ID from Microsoft's Marketplace?

Nope, that is not be allowed. It just adds support for the two workflows mentioned in the "How to test" section where in the "Install by ID" workflow the data source would be the registry the VSX extension is connected to, by default OpenVSX.org.

**Review by azatsarynnyy — COMMENTED** [review]
There a couple of new files added with the Red Hat copyright header.
@fgreinacher shouldn't it be your company's copyright?

**Comment by fgreinacher** [comment]
> There a couple of new files added with the Red Hat copyright header.
> @fgreinacher shouldn't it be your company's copyright?

Good hint, I'll add our copyright, and will also leave Red Hat's because I copied most parts.

**Review by azatsarynnyy — APPROVED** [review]
I've checked installing from both the local file and OpenVSX registry. The extensions have been installed successfully. Thanks!

File: packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts
Comment (benoitf) [inline packages/plugin-ext-vscode/src/browser/plugin-vscode-commands-contribution.ts:218]: we don't have anymore hooks for this protocol so what is happening ?

File: packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts
Comment (benoitf) [reply packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts:27]: You're giving
`local-file:${URI.revive(vsixUriOrExtensionId).fsPath}`

so it looks like a path no ?

And you can say that `file://` only supports file, not directories

**Comment by azatsarynnyy** [comment]
The tests are failed, but I'm not sure whether it's related to this PR.

File: packages/plugin-ext/src/main/node/resolvers/util.ts
Comment (benoitf) [inline packages/plugin-ext/src/main/node/resolvers/util.ts:25]: should we use inheritance between the deployers over custom functions ?

If rewritten, we can use fs-extra and await fs.pathExists instead of all `sync` methods

Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/util.ts]: Good suggestion, feels much better now: 46b653b9b78c362e4b5f637f8b81e0b7d0190ebc

**Review by benoitf — APPROVED** [review]

File: packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts.ts
Comment (DukeNgn) [inline packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts.ts:1]: The filename shouldn't be `.ts.ts`

File: packages/plugin-ext/src/main/node/plugin-ext-backend-module.ts
Comment (DukeNgn) [inline packages/plugin-ext/src/main/node/plugin-ext-backend-module.ts:28]: should be `from './resolvers/local-file-plugin-deployer-resolver'` ?

File: packages/plugin-ext/src/main/node/plugin-ext-backend-module.ts
Comment (DukeNgn) [inline packages/plugin-ext/src/main/node/plugin-ext-backend-module.ts:27]: The file was deleted. There is nowhere for this to import from.

I believe there's more instant of this file being imported. If you delete the file, please resolve these instants.

**Review by DukeNgn — CHANGES_REQUESTED** [review]
Have a couple of things I believe you should fix (the CI is failing because of the changes in this PR).

https://travis-ci.com/github/eclipse-theia/theia/jobs/443571087#L544-L556

File: packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts.ts
Comment (DukeNgn) [inline packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts.ts:1]: The file name shouldn't be `.ts.ts` as well

File: packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts
Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts]: Yes, it's a path, but a path is not an URI. They have different rules for encoding stuff and mixing them up usually leads to weird issues.

So I would rather prefer to stay with `local-file` because it makes it pretty clear that it has the same format as `local-dir`, just referring to files instead of directories. 

If necessary, we could still add support for the `file` scheme in a separate MR but this should then be taking real URIs and be able to handle both files and directories.

Would that be OK for you?

Comment (benoitf) [reply packages/plugin-ext/src/main/node/resolvers/plugin-local-file-resolver.ts]: yes I approved PR yesterday

**Comment by fgreinacher** [comment]
Thanks for your reviews! Not sure who is going to press the merge button here, but it seems that approvals from @vince-fugnitto, @tolusha and @DukeNgn are missing at the moment. Would be awesome to have this in 1.8.0 👍

**Comment by azatsarynnyy** [comment]
@fgreinacher no problem, let's wait for about 2 hours and I'll merge it today if there're no more comments.

**Review by vince-fugnitto — COMMENTED** [review]
I verified the functionality and it worked well for me :+1: 
- confirmed that the install works correctly for `id`
- confirmed that the install works correctly for `.vsix` paths

My only comment would be to add a [breaking change](https://github.com/eclipse-theia/theia/blob/master/doc/pull-requests.md#checklist-breaking-changes) note in the `changelog` as the pull-request includes the deletion of a file which extenders will likely need to address.

**Comment by azatsarynnyy** [comment]
@fgreinacher let me know when you fix the latest comment from @vince-fugnitto and I'll merge it.
Please, don't forget to squash the commits.

**Comment by fgreinacher** [comment]
@vince-fugnitto Good hint, I updated the changelog. 
@azatsarynnyy All done and ready for merge once CI is finished 👍🏽

**Review by vince-fugnitto — COMMENTED** [review]
@fgreinacher  please fix the failing CI: https://travis-ci.com/github/eclipse-theia/theia/jobs/444517542#L541-L543.

**Comment by fgreinacher** [comment]
CI does not seem to be triggered anymore. Any ideas @vince-fugnitto?

**Comment by vince-fugnitto** [comment]
> CI does not seem to be triggered anymore. Any ideas @vince-fugnitto?

I'm not sure, an easy fix would be to perform an `amend` and `force push`.

**Comment by vince-fugnitto** [comment]
> CI does not seem to be triggered anymore. Any ideas @vince-fugnitto?

I believe it is related to the changes by Travis: https://github.com/eclipse-theia/theia/pull/8765#issuecomment-731377516.

**Comment by fgreinacher** [comment]
> > CI does not seem to be triggered anymore. Any ideas @vince-fugnitto?
> 
> I believe it is related to the changes by Travis: [#8765 (comment)](https://github.com/eclipse-theia/theia/pull/8765#issuecomment-731377516).

😿

**Comment by azatsarynnyy** [comment]
> > CI does not seem to be triggered anymore. Any ideas @vince-fugnitto?
> 
> I'm not sure, an easy fix would be to perform an `amend` and `force push`.

@fgreinacher could you try this suggestion ^^ in order to retrigger the tests on Travis?

**Comment by DukeNgn** [comment]
> > > CI does not seem to be triggered anymore. Any ideas @vince-fugnitto?
> > 
> > 
> > I'm not sure, an easy fix would be to perform an `amend` and `force push`.
> 
> @fgreinacher could you try this suggestion ^^ in order to retrigger the tests on Travis?

We have run out of credits on Travis so the CI is temporarily disabled for all PRs at the moment.

File: CHANGELOG.md
Comment (vince-fugnitto) [inline CHANGELOG.md:7]:
```suggestion
<a name="breaking_changes_1.9.0">[Breaking Changes:](#breaking_changes_1.9.0)</a>
```

File: packages/plugin-ext/src/main/node/resolvers/local-directory-plugin-deployer-resolver.ts
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/node/resolvers/local-directory-plugin-deployer-resolver.ts:17]: Remove?

File: packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts
Comment (vince-fugnitto) [inline packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts:17]: Remove?

Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts]: Thanks, sure 👍

File: packages/plugin-ext/src/main/node/resolvers/local-directory-plugin-deployer-resolver.ts
Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/local-directory-plugin-deployer-resolver.ts]: Thanks, sure 👍

File: CHANGELOG.md
Comment (fgreinacher) [reply CHANGELOG.md]: Thanks, sure 👍

**Comment by fgreinacher** [comment]
Hm, one job seems to be stuck: https://github.com/eclipse-theia/theia/pull/8745/checks?check_run_id=1481816411

**Comment by azatsarynnyy** [comment]
> Hm, one job seems to be stuck: https://github.com/eclipse-theia/theia/pull/8745/checks?check_run_id=1481816411

I've rerun it

**Comment by azatsarynnyy** [comment]
> Hm, one job seems to be stuck: https://github.com/eclipse-theia/theia/pull/8745/checks?check_run_id=1481816411

I saw that check has failed on other PRs as well but for a different reason.
I'm not sure if can skip it to merge the PR.
cc @kittaakos @vince-fugnitto

**Review by DukeNgn — COMMENTED** [review]
> > Hm, one job seems to be stuck: [#8745 (checks)](https://github.com/eclipse-theia/theia/pull/8745/checks?check_run_id=1481816411)
> 
> I saw that check has failed on other PRs as well but for a different reason.
> I'm not sure if can skip it to merge the PR.
> cc @kittaakos @vince-fugnitto

```

@theia/example-browser: Error: The number of constructor arguments in the derived class LocalDirectoryPluginDeployerResolver must be >= than the number of constructor arguments of its base class.\n    at 
/home/runner/work/theia/theia/node_modules/inversify/lib/planning/planner.js:111:27\n    at Array.forEach (<anonymous>)\n    at _createSubRequests 
(/home/runner/work/theia/theia/node_modules/inversify/lib/planning/planner.js:94:20)\n    at 
/home/runner/work/theia/theia/node_modules/inversify/lib/planning/planner.js:115:17\n    at Array.forEach (<anonymous>)\n    at 
/home/runner/work/theia/theia/node_modules/inversify/lib/planning/planner.js:114:26\n    at Array.forEach (<anonymous>)\n    at _createSubRequests 
(/home/runner/work/theia/theia/node_modules/inversify/lib/planning/planner.js:94:20)\n    at 
/home/runner/work/theia/theia/node_modules/inversify/lib/planning/planner.js:115:17\n    at Array.forEach (<anonymous>)

```

https://github.com/eclipse-theia/theia/pull/8745/checks?check_run_id=1482104865#step:7:24


Looking at the log, I think the CI is failing because of the changes from this PR?

File: packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts
Comment (DukeNgn) [inline packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts:26]: `LocalFilePluginDeployerResolver` provides 0 arguments in the constructor while the base class `LocalPluginDeployerResolver` has 1 argument in its constructor.

I think this causes the problem in the CI.

Comment (vince-fugnitto) [reply packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts:26]: As mentioned, it is not related to CI, the check fails locally with the error: https://github.com/eclipse-theia/theia/pull/8745#pullrequestreview-542313320.
It can be reproduced by running:

```
npx run test @theia/example-browser
```

Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts:26]: Sorry folks, it seems I messed up something when rebasing this PR against the latest `master`. I actually already fixed these issues some time ago. I will take care and ping you again when I fixed this.

Comment (fgreinacher) [reply packages/plugin-ext/src/main/node/resolvers/local-file-plugin-deployer-resolver.ts]: @azatsarynnyy Thanks for your patience. CI is now green, and I also verified locally that Theia is starting properly.

**Review by azatsarynnyy — APPROVED** [review]
I've re-checked it again after the changes have been made.
Works as expected. Thanks!

**Pull request merged**
